litestar/plugins/problem_details.py

```diff
@@ -0,0 +1,111 @@
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Callable, Mapping, TypeVar
+
+from litestar.connection.request import Request
+from litestar.exceptions.http_exceptions import HTTPException
+from litestar.plugins.base import InitPluginProtocol
+from litestar.response.base import Response
+from litestar.serialization.msgspec_hooks import encode_json
+from litestar.types.callable_types import ExceptionHandler, ExceptionT
+
+if TYPE_CHECKING:
+    from litestar.config.app import AppConfig
+
+ProblemDetailsExceptionT = TypeVar("ProblemDetailsExceptionT", bound="ProblemDetailsException")
+ProblemDetailsExceptionHandlerType = Callable[[Request, ProblemDetailsExceptionT], Response]
+ExceptionToProblemDetailMapType = Mapping[type[ExceptionT], Callable[[ExceptionT], ProblemDetailsExceptionT]]
+
+
+def _problem_details_exception_handler(request: Request, exc: ProblemDetailsException) -> Response:
+    return exc.to_response(request)
+
+
+def _create_exception_handler(
+    exc_to_problem_details_exc_fn: Callable[[ExceptionT], ProblemDetailsException], exc_type: type[ExceptionT]
+) -> ExceptionHandler:
+    def _exception_handler(req: Request, exc: exc_type) -> Response:
+        problem_details_exc = exc_to_problem_details_exc_fn(exc)
+
+        return problem_details_exc.to_response(req)
+
+    return _exception_handler
+
+
+class ProblemDetailsException(HTTPException):
+    _PROBLEM_DETAIL_HEADER = "application/problem+json"
+
+    def __init__(
+        self,
+        type_: str = "about:blank",
+        status_code: int | None = None,
+        title: str | None = None,
+        instance: str | None = None,
+        detail: str | dict[str, Any] | list[Any] | None = None,
+        extra: dict[str, Any] | None = None,
+        headers: dict[str, Any] | None = None,
+    ) -> None:
+        self.type_ = type_
+        self.status_code = status_code or self.status_code
+        self.title = title
+        self.instance = instance
+        self.detail = detail  # type: ignore
+        self.extra = extra
+        self.headers = headers
+
+    def to_response(self, request: Request) -> Response:
+        problem_details = {"type": self.type_, "status": self.status_code}
+        if self.title is not None:
+            problem_details["title"] = self.title
+        if self.instance is not None:
+            problem_details["instance"] = self.instance
+        if self.detail is not None:
+            problem_details["detail"] = self.detail
+        if self.extra is not None:
+            problem_details.update(self.extra)
+
+        content = encode_json(problem_details, request.route_handler.default_serializer)
+        headers = self.headers or {}
+        headers["content-type"] = self._PROBLEM_DETAIL_HEADER
+
+        return Response(content, headers=headers)
+
+    @classmethod
+    def from_http_exception(cls, exc: HTTPException) -> ProblemDetailsException:
+        return ProblemDetailsException(status_code=exc.status_code)
```

Comment:
I don't quite see the point of converting HTTPException to ProblemDetailsException? I feel like what we should have is
1) ProblemDetailsException
2) Exception handler that catches ProblemDetailsException and makes it into an application/problem+json response
3) Exception handler that catches HTTPException and makes it into an application/problem+json response

1 and 2 work on their own, whereas 3 is a change to default behavior

Reply:
Also, isn't there more that can be salvaged from an HTTPException than just the status code? Don't the standard litestar error responses already have a "detail" entry for example?

Reply:
I'd kept it there in the beginning, but I would probably move that into a separate function rather than as a classmethod. There's no real benefit since users overriding it provide no benefit either.

Reply:
> Also, isn't there more that can be salvaged from an HTTPException than just the status code? Don't the standard litestar error responses already have a "detail" entry for example?

Yes, there should be more that we can extract from the `HTTPException` such as `detail` and/or `extra`.

Reply:
> Edit: Is it just to enable the exception_to_problem_detail_map feature? Is that worth it though? It will only handle conversion from one type of exception (barring some aggressive use of isinstance) anyway, right?

I didn't quite understand. It's for cases where you want to handle a specific exception in a specific manner. For example, you might want to convert all `ValidationError` exceptions into a specific format whereas you might want to handle another exception differently. If this wasn't provided, then users have no control over the transformation of the exception into a problem detail. It's the same way litestar takes in the custom exception handlers, but instead of returning a `Response`, you return a `ProblemDetailException.`

Also, by providing this, users can override the handling of `HTTPException` as well by just providing a callable that does the transformation.